# EDA básico datos del proyecto
## Se realiza un análisis exploratorio de datos básico de los datos para entender los campos, sus valores y sus relaciones.

___

### Importaciones

#### Librerías:

In [9]:
# Importaciones de librerías
import pandas as pd

#### Datos:

In [10]:
# Importaciones de datos
df_customers = pd.read_csv('../../Data/CUSTOMERS.csv')
df_geolocation = pd.read_csv('../../Data/GEO_LOCATION.csv')
df_products = pd.read_csv('../../Data/PRODUCTS.csv')
df_sellers = pd.read_csv('../../Data/SELLERS.csv')
df_orders = pd.read_csv('../../Data/ORDERS.csv')
df_order_items = pd.read_csv('../../Data/ORDER_ITEMS.csv')
df_order_payments = pd.read_csv('../../Data/ORDER_PAYMENTS.csv')
df_order_review = pd.read_csv('../../Data/ORDER_REVIEW_RATINGS.csv')

___

### Funciones

In [11]:
### FUNCIÓN 1: Crea un reporte de la cardinalidad y saca los datos como df ####

def reporte_cardinalidad(df):
    # Calculamos el total de filas una sola vez
    total_filas = len(df)
    
    # Creamos un DataFrame con los resultados
    reporte = pd.DataFrame({
        'Columna': df.columns,
        'Nunique': [df[col].nunique() for col in df.columns],
        'Cardinalidad_%': [(df[col].nunique() / total_filas) * 100 for col in df.columns],
        'Tipo': [df[col].dtype for col in df.columns]
    })

    reporte['Cardinalidad_%'] = reporte['Cardinalidad_%'].round(2)
    # Añadimos una recomendación rápida basada en el %
    def sugerir_tipo(row):
        if row['Cardinalidad_%'] == 100:
            return "Posible ID / Clave Primaria"
        elif row['Cardinalidad_%'] < 5:
            return "Categórica (Baja cardinalidad)"
        else:
            return "Numérica / Categórica compleja"
            
    reporte['Sugerencia'] = reporte.apply(sugerir_tipo, axis=1)
    
    return reporte.sort_values(by='Cardinalidad_%', ascending=False)



### FUNCIÓN 2: Obtiene las fechas máximas y mínimas de cada campo de cada tabla #####

def resumen_fechas(tablas: dict):
    """
    Recorre un diccionario de DataFrames y devuelve un DataFrame resumen
    con las columnas de tipo datetime y sus valores mínimo y máximo.
    
    Parámetros:
        tablas (dict): {"nombre_tabla": dataframe}
    
    Retorna:
        DataFrame con columnas:
        - tabla
        - campo
        - fecha_min
        - fecha_max
    """
    
    registros = []

    for nombre_tabla, df in tablas.items():
        # Recorremos columnas del DataFrame
        for columna in df.columns:
            # Detectamos columnas datetime
            if pd.api.types.is_datetime64_any_dtype(df[columna]):
                registros.append({
                    "tabla": nombre_tabla,
                    "campo": columna,
                    "fecha_min": df[columna].min(),
                    "fecha_max": df[columna].max()
                })

    return pd.DataFrame(registros)


___

## Resumen

In [12]:
# Resumen de filas por tabla
resumen = {
    "Tabla": ["customers", "geolocation", "products", "sellers", "orders", "order_items", "order_payments","order_reviews"],
    "Filas": [
        len(df_customers),
        len(df_geolocation),
        len(df_products),
        len(df_sellers),
        len(df_orders),
        len(df_order_items),
        len(df_order_payments),
        len(df_order_review)
    ],
    "Columnas": [
        df_customers.shape[1],
        df_geolocation.shape[1],
        df_products.shape[1],
        df_sellers.shape[1],
        df_orders.shape[1],
        df_order_items.shape[1],
        df_order_payments.shape[1],
        df_order_review.shape[1]
    ]
}

df_resumen = pd.DataFrame(resumen)
df_resumen


,Tabla,Filas,Columnas
0,customers,99441,5
1,geolocation,19015,5
2,products,32327,9
3,sellers,3095,4
4,orders,96455,8
5,order_items,112650,7
6,order_payments,103886,5
7,order_reviews,100000,5


## Análisis de las tablas

___

### Datos tabla CUSTOMERS.csv

In [13]:
df_customers.shape

(99441, 5)

In [14]:
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


In [15]:
df_customers.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,Adilabad,Andhra Pradesh
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,Adoni,Andhra Pradesh
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,Akkarampalle,Andhra Pradesh
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,Akkayapalle,Andhra Pradesh
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,Alwal,Andhra Pradesh


In [16]:
df_customers.describe(include='all')

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
count,99441,99441,99441.000000,99441,99441
unique,99441,96096,NaN,4119,20
top,06b8999e2fba1a1fbc88172c00ba8bc7,8d50f5eadf50201ccdcedfb9e2ac8455,NaN,Akkarampalle,Andhra Pradesh
freq,1,17,NaN,15540,60688
mean,NaN,NaN,35137.474583,NaN,NaN
std,NaN,NaN,29797.938996,NaN,NaN
min,NaN,NaN,1003.000000,NaN,NaN
25%,NaN,NaN,11347.000000,NaN,NaN
50%,NaN,NaN,24416.000000,NaN,NaN
75%,NaN,NaN,58900.000000,NaN,NaN


In [17]:
reporte_cardinalidad(df_customers)


,Columna,Nunique,Cardinalidad_%,Tipo,Sugerencia
0,customer_id,99441,100.00,object,Posible ID / Clave Primaria
1,customer_unique_id,96096,96.64,object,Numérica / Categórica compleja
2,customer_zip_code_prefix,14994,15.08,int64,Numérica / Categórica compleja
3,customer_city,4119,4.14,object,Categórica (Baja cardinalidad)
4,customer_state,20,0.02,object,Categórica (Baja cardinalidad)


____

### Datos tabla GEO_LOCATION.csv

In [18]:
df_geolocation.shape

(19015, 5)

In [19]:
df_geolocation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19015 entries, 0 to 19014
Data columns (total 5 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   geolocation_zip_code_prefix  19015 non-null  int64  
 1   geolocation_lat              19015 non-null  float64
 2   geolocation_lng              19015 non-null  float64
 3   geolocation_city             19015 non-null  object 
 4   geolocation_state            19015 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 742.9+ KB


In [20]:
df_geolocation.head()

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,Akkarampalle,Andhra Pradesh
1,1046,-23.546081,-46.644820,Akkarampalle,Andhra Pradesh
2,1041,-23.544392,-46.639499,Akkarampalle,Andhra Pradesh
3,1035,-23.541578,-46.641607,Akkarampalle,Andhra Pradesh
4,1012,-23.547762,-46.635361,Akkarampalle,Andhra Pradesh


In [21]:
df_geolocation.describe(include='all')

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
count,19015.000000,19015.000000,19015.000000,19015,19015
unique,NaN,NaN,NaN,3809,20
top,NaN,NaN,NaN,Akkarampalle,Andhra Pradesh
freq,NaN,NaN,NaN,5889,11218
mean,42711.591901,-19.062087,-46.058008,NaN,NaN
std,30905.051745,7.319402,5.380751,NaN,NaN
min,1001.000000,-36.605374,-72.927296,NaN,NaN
25%,12721.500000,-23.564386,-49.000445,NaN,NaN
50%,38240.000000,-22.429252,-46.632544,NaN,NaN
75%,70656.500000,-15.615448,-43.255324,NaN,NaN


In [22]:
reporte_cardinalidad(df_geolocation)

,Columna,Nunique,Cardinalidad_%,Tipo,Sugerencia
0,geolocation_zip_code_prefix,19015,100.00,int64,Posible ID / Clave Primaria
2,geolocation_lng,18988,99.86,float64,Numérica / Categórica compleja
1,geolocation_lat,18987,99.85,float64,Numérica / Categórica compleja
3,geolocation_city,3809,20.03,object,Numérica / Categórica compleja
4,geolocation_state,20,0.11,object,Categórica (Baja cardinalidad)


___

### Datos tabla PRODUCTS.csv

In [23]:
df_products.shape

(32327, 9)

In [24]:
df_products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32327 entries, 0 to 32326
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   product_id                  32327 non-null  object
 1   product_category_name       32327 non-null  object
 2   product_name_lenght         32327 non-null  int64 
 3   product_description_lenght  32327 non-null  int64 
 4   product_photos_qty          32327 non-null  int64 
 5   product_weight_g            32327 non-null  int64 
 6   product_length_cm           32327 non-null  int64 
 7   product_height_cm           32327 non-null  int64 
 8   product_width_cm            32327 non-null  int64 
dtypes: int64(7), object(2)
memory usage: 2.2+ MB


In [25]:
df_products.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,Perfumery,40,287,1,225,16,10,14
1,3aa071139cb16b67ca9e5dea641aaa2f,Art,44,276,1,1000,30,18,20
2,96bd76ec8810374ed1b65e291975717f,Sports_Leisure,46,250,1,154,18,9,15
3,cef67bcfe19066a932b7673e239eb23d,Baby,27,261,1,371,26,4,26
4,9dc1a7de274444849c219cff195d0b71,Housewares,37,402,4,625,20,17,13


In [26]:
df_products.describe(include='all')

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
count,32327,32327,32327.000000,32327.000000,32327.000000,32327.000000,32327.000000,32327.000000,32327.000000
unique,32327,71,NaN,NaN,NaN,NaN,NaN,NaN,NaN
top,1e9e8ef04dbcff4541ed26657ea517e5,Bed_Bath_Table,NaN,NaN,NaN,NaN,NaN,NaN,NaN
freq,1,3029,NaN,NaN,NaN,NaN,NaN,NaN,NaN
mean,NaN,NaN,48.473722,771.517277,2.188790,2276.960807,30.856498,16.955950,23.208464
std,NaN,NaN,10.246346,635.189674,1.736767,4279.734063,16.958460,13.637246,12.080665
min,NaN,NaN,5.000000,4.000000,1.000000,0.000000,7.000000,2.000000,6.000000
25%,NaN,NaN,42.000000,339.000000,1.000000,300.000000,18.000000,8.000000,15.000000
50%,NaN,NaN,51.000000,595.000000,1.000000,700.000000,25.000000,13.000000,20.000000
75%,NaN,NaN,57.000000,972.000000,3.000000,1900.000000,38.000000,20.500000,30.000000


In [27]:
reporte_cardinalidad(df_products)

,Columna,Nunique,Cardinalidad_%,Tipo,Sugerencia
0,product_id,32327,100.00,object,Posible ID / Clave Primaria
3,product_description_lenght,2960,9.16,int64,Numérica / Categórica compleja
5,product_weight_g,2201,6.81,int64,Numérica / Categórica compleja
7,product_height_cm,102,0.32,int64,Categórica (Baja cardinalidad)
6,product_length_cm,99,0.31,int64,Categórica (Baja cardinalidad)
8,product_width_cm,95,0.29,int64,Categórica (Baja cardinalidad)
1,product_category_name,71,0.22,object,Categórica (Baja cardinalidad)
2,product_name_lenght,66,0.20,int64,Categórica (Baja cardinalidad)
4,product_photos_qty,19,0.06,int64,Categórica (Baja cardinalidad)


___

### Datos tabla SELLERS.csv

In [28]:
df_sellers.shape

(3095, 4)

In [29]:
df_sellers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3038 non-null   object
 3   seller_state            3038 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB


In [30]:
df_sellers.head()

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,Alwal,Andhra Pradesh
1,d1b65fc7debc3361ea86b5f14c68d2e2,13023,Alwal,Andhra Pradesh
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,Badepalle,Andhra Pradesh
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,Akkarampalle,Andhra Pradesh
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,Koratla,Andhra Pradesh


In [31]:
df_sellers.describe(include='all')

,seller_id,seller_zip_code_prefix,seller_city,seller_state
count,3095,3095.000000,3038,3038
unique,3095,NaN,534,19
top,3442f8959a84dea7ee197c632cb2df15,NaN,Akkarampalle,Andhra Pradesh
freq,1,NaN,699,2060
mean,NaN,32174.790630,NaN,NaN
std,NaN,32656.670163,NaN,NaN
min,NaN,1001.000000,NaN,NaN
25%,NaN,7084.000000,NaN,NaN
50%,NaN,14940.000000,NaN,NaN
75%,NaN,61780.000000,NaN,NaN


In [32]:
reporte_cardinalidad(df_products)

,Columna,Nunique,Cardinalidad_%,Tipo,Sugerencia
0,product_id,32327,100.00,object,Posible ID / Clave Primaria
3,product_description_lenght,2960,9.16,int64,Numérica / Categórica compleja
5,product_weight_g,2201,6.81,int64,Numérica / Categórica compleja
7,product_height_cm,102,0.32,int64,Categórica (Baja cardinalidad)
6,product_length_cm,99,0.31,int64,Categórica (Baja cardinalidad)
8,product_width_cm,95,0.29,int64,Categórica (Baja cardinalidad)
1,product_category_name,71,0.22,object,Categórica (Baja cardinalidad)
2,product_name_lenght,66,0.20,int64,Categórica (Baja cardinalidad)
4,product_photos_qty,19,0.06,int64,Categórica (Baja cardinalidad)


___

### Datos tabla ORDERS.csv

In [33]:
df_orders.shape

(96455, 8)

In [34]:
df_orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96455 entries, 0 to 96454
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       96455 non-null  object
 1   customer_id                    96455 non-null  object
 2   order_status                   96455 non-null  object
 3   order_purchase_timestamp       96455 non-null  object
 4   order_approved_at              96455 non-null  object
 5   order_delivered_carrier_date   96455 non-null  object
 6   order_delivered_customer_date  96455 non-null  object
 7   order_estimated_delivery_date  96455 non-null  object
dtypes: object(8)
memory usage: 5.9+ MB


In [35]:
df_orders.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,10/2/2017 10:56,10/2/2017 11:07,10/4/2017 19:55,10/10/2017 21:25,10/18/2017 0:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,7/24/2018 20:41,7/26/2018 3:24,7/26/2018 14:31,8/7/2018 15:27,8/13/2018 0:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,8/8/2018 8:38,8/8/2018 8:55,8/8/2018 13:50,8/17/2018 18:06,9/4/2018 0:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,11/18/2017 19:28,11/18/2017 19:45,11/22/2017 13:39,12/2/2017 0:28,12/15/2017 0:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2/13/2018 21:18,2/13/2018 22:20,2/14/2018 19:46,2/16/2018 18:17,2/26/2018 0:00


In [36]:
df_orders['order_purchase_timestamp']=pd.to_datetime(df_orders['order_purchase_timestamp'])
df_orders['order_approved_at']=pd.to_datetime(df_orders['order_approved_at'])
df_orders['order_delivered_carrier_date']= pd.to_datetime(df_orders['order_delivered_carrier_date'])
df_orders['order_delivered_customer_date']=pd.to_datetime(df_orders['order_delivered_customer_date'])
df_orders['order_estimated_delivery_date']= pd.to_datetime(df_orders['order_estimated_delivery_date'])

In [37]:
df_orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96455 entries, 0 to 96454
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       96455 non-null  object        
 1   customer_id                    96455 non-null  object        
 2   order_status                   96455 non-null  object        
 3   order_purchase_timestamp       96455 non-null  datetime64[ns]
 4   order_approved_at              96455 non-null  datetime64[ns]
 5   order_delivered_carrier_date   96455 non-null  datetime64[ns]
 6   order_delivered_customer_date  96455 non-null  datetime64[ns]
 7   order_estimated_delivery_date  96455 non-null  datetime64[ns]
dtypes: datetime64[ns](5), object(3)
memory usage: 5.9+ MB


In [38]:
df_orders.describe(include='all')

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
count,96455,96455,96455,96455,96455,96455,96455,96455
unique,96455,96455,1,NaN,NaN,NaN,NaN,NaN
top,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,NaN,NaN,NaN,NaN,NaN
freq,1,1,96455,NaN,NaN,NaN,NaN,NaN
mean,NaN,NaN,NaN,2018-01-02 00:25:54.122855168,2018-01-02 10:42:36.795604224,2018-01-05 05:53:18.362344960,2018-01-14 13:48:55.026074368,2018-01-25 18:04:03.421284864
min,NaN,NaN,NaN,2016-09-15 12:16:00,2016-09-15 12:16:00,2016-10-08 10:34:00,2016-10-11 13:46:00,2016-10-04 00:00:00
25%,NaN,NaN,NaN,2017-09-14 09:38:30,2017-09-14 14:42:00,2017-09-18 17:06:00,2017-09-25 22:56:30,2017-10-05 00:00:00
50%,NaN,NaN,NaN,2018-01-20 20:00:00,2018-01-22 13:49:00,2018-01-24 16:28:00,2018-02-02 19:52:00,2018-02-16 00:00:00
75%,NaN,NaN,NaN,2018-05-05 18:53:30,2018-05-06 10:52:30,2018-05-08 14:34:30,2018-05-15 23:08:30,2018-05-28 00:00:00
max,NaN,NaN,NaN,2018-08-29 15:00:00,2018-08-29 15:10:00,2018-09-11 19:48:00,2018-10-17 13:22:00,2018-10-25 00:00:00


In [39]:
reporte_cardinalidad(df_orders)

,Columna,Nunique,Cardinalidad_%,Tipo,Sugerencia
0,order_id,96455,100.00,object,Posible ID / Clave Primaria
1,customer_id,96455,100.00,object,Posible ID / Clave Primaria
3,order_purchase_timestamp,86336,89.51,datetime64[ns],Numérica / Categórica compleja
6,order_delivered_customer_date,75633,78.41,datetime64[ns],Numérica / Categórica compleja
5,order_delivered_carrier_date,60999,63.24,datetime64[ns],Numérica / Categórica compleja
4,order_approved_at,49445,51.26,datetime64[ns],Numérica / Categórica compleja
7,order_estimated_delivery_date,445,0.46,datetime64[ns],Categórica (Baja cardinalidad)
2,order_status,1,0.00,object,Categórica (Baja cardinalidad)


___

### Datos tabla ORDER_ITEMS.csv

In [40]:
df_order_items.shape

(112650, 7)

In [41]:
df_order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [42]:
df_order_items.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,9/19/2017 9:45,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,5/3/2017 11:05,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,1/18/2018 14:48,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,8/15/2018 10:10,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2/13/2017 13:57,199.90,18.14


In [43]:
# Cambio el tipo de dato de `shipping_limit_date`
df_order_items['shipping_limit_date']=pd.to_datetime(df_order_items['shipping_limit_date'])

In [44]:
df_order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   order_id             112650 non-null  object        
 1   order_item_id        112650 non-null  int64         
 2   product_id           112650 non-null  object        
 3   seller_id            112650 non-null  object        
 4   shipping_limit_date  112650 non-null  datetime64[ns]
 5   price                112650 non-null  float64       
 6   freight_value        112650 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(1), object(3)
memory usage: 6.0+ MB


In [45]:
df_order_items.describe(include='all')

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
count,112650,112650.000000,112650,112650,112650,112650.000000,112650.000000
unique,98666,NaN,32951,3095,NaN,NaN,NaN
top,8272b63d03f5f79c56e9e4120aec44ef,NaN,aca2eb7d00ea1a7b8ebd4e68314663af,6560211a19b47992c3666cc44a7e94c0,NaN,NaN,NaN
freq,21,NaN,527,2033,NaN,NaN,NaN
mean,NaN,1.197834,NaN,NaN,2018-01-07 15:36:25.277763072,120.653739,19.990320
min,NaN,1.000000,NaN,NaN,2016-09-19 00:15:00,0.850000,0.000000
25%,NaN,1.000000,NaN,NaN,2017-09-20 20:57:15,39.900000,13.080000
50%,NaN,1.000000,NaN,NaN,2018-01-26 13:59:00,74.990000,16.260000
75%,NaN,1.000000,NaN,NaN,2018-05-10 14:33:45,134.900000,21.150000
max,NaN,21.000000,NaN,NaN,2020-04-09 22:35:00,6735.000000,409.680000


In [46]:
reporte_cardinalidad(df_order_items)

,Columna,Nunique,Cardinalidad_%,Tipo,Sugerencia
0,order_id,98666,87.59,object,Numérica / Categórica compleja
4,shipping_limit_date,54615,48.48,datetime64[ns],Numérica / Categórica compleja
2,product_id,32951,29.25,object,Numérica / Categórica compleja
6,freight_value,6999,6.21,float64,Numérica / Categórica compleja
5,price,5968,5.30,float64,Numérica / Categórica compleja
3,seller_id,3095,2.75,object,Categórica (Baja cardinalidad)
1,order_item_id,21,0.02,int64,Categórica (Baja cardinalidad)


___

### Datos tabla ORDER_PAYMENTS.csv

In [47]:
df_order_payments.shape

(103886, 5)

In [48]:
df_order_payments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB


In [49]:
df_order_payments.head()

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


In [50]:
df_order_payments.describe(include='all')

,order_id,payment_sequential,payment_type,payment_installments,payment_value
count,103886,103886.000000,103886,103886.000000,103886.000000
unique,99440,NaN,5,NaN,NaN
top,fa65dad1b0e818e3ccc5cb0e39231352,NaN,credit_card,NaN,NaN
freq,29,NaN,76795,NaN,NaN
mean,NaN,1.092679,NaN,2.853349,154.100380
std,NaN,0.706584,NaN,2.687051,217.494064
min,NaN,1.000000,NaN,0.000000,0.000000
25%,NaN,1.000000,NaN,1.000000,56.790000
50%,NaN,1.000000,NaN,1.000000,100.000000
75%,NaN,1.000000,NaN,4.000000,171.837500


In [51]:
reporte_cardinalidad(df_order_payments)

,Columna,Nunique,Cardinalidad_%,Tipo,Sugerencia
0,order_id,99440,95.72,object,Numérica / Categórica compleja
4,payment_value,29077,27.99,float64,Numérica / Categórica compleja
1,payment_sequential,29,0.03,int64,Categórica (Baja cardinalidad)
3,payment_installments,24,0.02,int64,Categórica (Baja cardinalidad)
2,payment_type,5,0.00,object,Categórica (Baja cardinalidad)


___

### Datos tabla ORDER_REVIEW_RATINGS.csv

In [52]:
df_order_review.shape

(100000, 5)

In [53]:
df_order_review.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column                   Non-Null Count   Dtype 
---  ------                   --------------   ----- 
 0   review_id                100000 non-null  object
 1   order_id                 100000 non-null  object
 2   review_score             100000 non-null  int64 
 3   review_creation_date     100000 non-null  object
 4   review_answer_timestamp  100000 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


In [54]:
df_order_review.head()

,review_id,order_id,review_score,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,1/18/2018 0:00,1/18/2018 21:46
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,3/10/2018 0:00,3/11/2018 3:05
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,2/17/2018 0:00,2/18/2018 14:36
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,4/21/2017 0:00,4/21/2017 22:02
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,3/1/2018 0:00,3/2/2018 10:26


In [55]:
# Cambio tipos de datos `review_creation_date` y `review_answer_timestamp`a `datetime`
df_order_review['review_creation_date']=pd.to_datetime(df_order_review['review_creation_date'])
df_order_review['review_answer_timestamp']= pd.to_datetime(df_order_review['review_answer_timestamp'])

In [56]:
df_order_review.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   review_id                100000 non-null  object        
 1   order_id                 100000 non-null  object        
 2   review_score             100000 non-null  int64         
 3   review_creation_date     100000 non-null  datetime64[ns]
 4   review_answer_timestamp  100000 non-null  datetime64[ns]
dtypes: datetime64[ns](2), int64(1), object(2)
memory usage: 3.8+ MB


In [108]:
df_order_review.describe(include='all')

,review_id,order_id,review_score,review_creation_date,review_answer_timestamp
count,100000,100000,100000.000000,100000,100000
unique,99173,99441,NaN,NaN,NaN
top,2172867fd5b1a55f98fe4608e1547b4b,03c939fd7fd3b38f8485a0f95798f1f6,NaN,NaN,NaN
freq,3,3,NaN,NaN,NaN
mean,NaN,NaN,4.070890,2018-01-12 17:58:10.739999744,2018-01-15 21:30:16.925400320
min,NaN,NaN,1.000000,2016-10-02 00:00:00,2016-10-07 18:32:00
25%,NaN,NaN,4.000000,2017-09-23 00:00:00,2017-09-27 01:19:30
50%,NaN,NaN,5.000000,2018-02-02 00:00:00,2018-02-04 19:30:30
75%,NaN,NaN,5.000000,2018-05-15 00:00:00,2018-05-20 10:59:45
max,NaN,NaN,5.000000,2018-08-31 00:00:00,2018-10-29 12:27:00


In [58]:
reporte_cardinalidad(df_order_review)

,Columna,Nunique,Cardinalidad_%,Tipo,Sugerencia
1,order_id,99441,99.44,object,Numérica / Categórica compleja
0,review_id,99173,99.17,object,Numérica / Categórica compleja
4,review_answer_timestamp,89500,89.50,datetime64[ns],Numérica / Categórica compleja
3,review_creation_date,637,0.64,datetime64[ns],Categórica (Baja cardinalidad)
2,review_score,5,0.00,int64,Categórica (Baja cardinalidad)


____

## Conclusiones iniciales

### SOBRE LAS TABLAS

___

Tenemos 8 tablas de las cuales podemos inferir que:  
* Exiten 4 tablas (`CUSTOMERS.csv`, `GEO_LOCATION.csv`,`PRODUCTS.csv`y `ORDERS.csv`) como tablas de dimensiones con entidades fuertes
* La tabla `ORDERS.csv` que es la tabla de hechos central donde se unen los pedidos y clientes. Sería una entidad débil por parte de la tabla `CUSTOMERS.csv`; pero a la vez es una entidad fuerte para las tablas `ORDER_ITEMS.csv`,`ORDER_PAYMENTS.csv`y `ORDER_REVIEW_RATINGS.csv`.
* El resto de tablas (`ORDER_ITEMS.csv`,`ORDER_PAYMENTS.csv`y `ORDER_REVIEW_RATINGS.csv`) no tienen existencia propia sin la tabla `ORDERS.CSV`.

TABLAS DE DIMENSIONES:

| Tabla | Entidad | Tipo de entidad | ID | DESCRIPCIÓN| OBSERVACIONES|
|:-----:|:-------:|:---------------:|:--:|:----------:|:------------:|
|CUSTOMER.csv| cliente | Fuerte | customer_id | Información sobre clientes | Tiene otro campo 'customer_unique_id'. Al parecer hace referencia al cliente. El customer_id se genera por pedido|
|GEO_LOCATION.csv| localizacion | Fuerte | geolocation_zip_code_prefix | Código postal, coordenadas y ciudades ||
|PRODUCTS.csv| productos | Fuerte| product_id | Información sobre productos | Tiene campos que no tienen sentido |
|SELLERS.csv| vendedores| Fuerte | seller_id | Información sobre vendedores | Añaden campos de localización|

TABLA DE HECHOS PRINCIPAL:

| Tabla | Entidad | Tipo de entidad | ID | DESCRIPCIÓN| OBSERVACIONES|
|:-----:|:-------:|:---------------:|:--:|:----------:|:------------:|
| ORDERS.csv| pedidos | Débil/Fuerte | order_id (PK), customer_id (candidata)| Información sobre pedidos | Por cada pedido, se genera un customer_id que luego se relaciona con los clientes únicos|

TABLAS DEPENDIENTES DE ORDERS.csv:

| Tabla | Entidad | Tipo de entidad | ID | DESCRIPCIÓN| OBSERVACIONES|
|:-----:|:-------:|:---------------:|:--:|:----------:|:------------:|
|ORDER_ITEMS.csv| productosXpedido|Débil existencia e identificación (ORDERS.csv)| Compuesta por order_id y order_item_id| Información de productos por pedido | Aparece el precio del producto y costes de entrega además de contener los id de customer_id y seller_id.
|ORDER_PAYMENTS.csv| pagos | Débil existencia e identificación (ORDERS.csv)| Compuesta por order_id y payment_sequential| Información sobre pagos y plazos | Recoge el importe total del precio del producto más el coste de envío|
|ORDER_REVIEW_RATINGS.csv| reviews| Débil existencia e identificación (ORDERS.csv)| Compuesta por review_id y order_id| Valoraciones de los clientes| Información sobre cuándo se realizaron| 

### SOBRE VALORES

* No existen valores nulos

DISCREPANCIAS ENTRE CANTIDAD DE VALORES:
* La tabla **ORDERS.csv** tiene 96.455 registros únicos, mientras que la tabla **CUSTOMERS.csv** en su campo `customer_unique_id` solo tiene 96.096; por lo que contiene más clientes de los que se tiene referencia en dicha tabla.
* La tabla **ORDER_ITEMS.csv** cotiene 98.666 registros únicos. Se deduce que son un desglose por producto de los pedidos reealizados con el precio de dicho producto y los costes de envío. En la tabla **ORDER_PAYMENTS.csv** hay 99.440 registros únicos, por lo que hay más pagos que productos comprados por pedido.
* Así mismo, la tabla **ORDER_ITEMS.csv** tiene 32.951 productos únicos, sin embargo en la tabla **PRODUCTS.csv**, el total de prodcutos únicos es de 32.327, por lo que hay productos vendidos que no se recogen en la tabla principal de productos.
* El número de vendedores coincide entre los registros únicos de la tabla **SELERS.csv** y el `seller_id` de la tabla **ORDER_ITEMS.csv**

Datos resumidos:

| Tabla               | Campo                | Valores únicos | Cardinalidad |
|---------------------|----------------------|----------------|--------------|
| ORDERS.csv          | order_id             | 96.455         | 100,00%      |
|                     | customer_id          | 96.455         | 100,00%      |
| ORDER_ITEMS.csv     | order_id             | 98.666         | 87,59%       |
|                     | product_id           | 32.951         | 29,25%       |
|                     | seller_id            | 3.095          | 2,75%        |
| ORDER_PAYMENT.csv   | order_id             | 99.440         | 95,72%       |
| ORDER_REVIEW_RATING | review_id            | 99.173         | 99,44%       |
|                     | order_id             | 99.441         | 99,17%       |
| CUSTOMERS.csv       | customer_id          | 99.441         | 100,00%      |
|                     | customer_unique_id   | 96.096         | 96,64%       |
| PRODUCTS.csv        | product_id           | 32.327         | 100,00%      |
| SELLERS.csv         | seller_id            | 3.095          | 100,00%      |


### SOBRE FECHAS

In [59]:
# Defino el diccionario con las tablas
tablas = {
    "customers": df_customers,
    "geolocation": df_geolocation,
    "products": df_products,
    "sellers": df_sellers,
    "orders": df_orders,
    "order_items": df_order_items,
    "order_payments": df_order_payments,
    "order_review": df_order_review
   
}

# Ejecuto la función
df_rangos_fechas = resumen_fechas(tablas)
df_rangos_fechas

,tabla,campo,fecha_min,fecha_max
0,orders,order_purchase_timestamp,2016-09-15 12:16:00,2018-08-29 15:00:00
1,orders,order_approved_at,2016-09-15 12:16:00,2018-08-29 15:10:00
2,orders,order_delivered_carrier_date,2016-10-08 10:34:00,2018-09-11 19:48:00
3,orders,order_delivered_customer_date,2016-10-11 13:46:00,2018-10-17 13:22:00
4,orders,order_estimated_delivery_date,2016-10-04 00:00:00,2018-10-25 00:00:00
5,order_items,shipping_limit_date,2016-09-19 00:15:00,2020-04-09 22:35:00
6,order_review,review_creation_date,2016-10-02 00:00:00,2018-08-31 00:00:00
7,order_review,review_answer_timestamp,2016-10-07 18:32:00,2018-10-29 12:27:00


In [60]:
fecha_minima= df_rangos_fechas['fecha_min'].min()
fecha_maxima= df_rangos_fechas['fecha_max'].max()
print(f"La fecha mínima es {fecha_minima}")
print(f"La fecha máxima es {fecha_maxima}")

La fecha mínima es 2016-09-15 12:16:00
La fecha máxima es 2020-04-09 22:35:00


____

## Comprobación integridad tablas

### CUSTOMERS <-> ORDERS via `customer_id`

Se revisa la discrepancia encontrada entre la tabla **ORDERS.csv** y la tabla **CUSTOMER.csv** a través de la clave primaria/foránea `customer_id`:

In [61]:
# 1. Lista de customer_id que aparecen en orders
clientes_con_pedido = set(df_orders['customer_id'])

# 2. Clientes que NO aparecen en orders
clientes_sin_pedido = df_customers[~df_customers['customer_id'].isin(clientes_con_pedido)]

# 3. Lista de customer_id válidos (los de customers)
clientes_validos = set(df_customers['customer_id'])

# 4. Pedidos cuyo customer_id NO existe en customers
orders_invalidos = df_orders[~df_orders['customer_id'].isin(clientes_validos)]

# Resultados
print("Valores totales de clientes en tabla CUSTOMERS:", (df_customers['customer_id'].shape[0]))
print("Clientes únicos  en tabla CUSTOMERS ('customer_unique_id'):", len(df_customers['customer_unique_id'].unique()))
print("Clientes en tabla ORDERS: ",len(clientes_con_pedido))
print(f"Clientes en tabla CUSTOMERS sin pedidos en ORDERS': {clientes_sin_pedido.shape[0]}")
print(f"Clientes en ORDERS que no están en CUSTOMERS: {orders_invalidos.shape[0]}")


Valores totales de clientes en tabla CUSTOMERS: 99441
Clientes únicos  en tabla CUSTOMERS ('customer_unique_id'): 96096
Clientes en tabla ORDERS:  96455
Clientes en tabla CUSTOMERS sin pedidos en ORDERS': 2986
Clientes en ORDERS que no están en CUSTOMERS: 0


### ORDER <-> ORDER_ITEM

In [106]:
# 1. Lista de pedidos
lista_pedidos = set(df_orders['order_id'])

# 2. Lista de pedidos que no se encuentra en order_items
pedidos_sin_productos =df_order_items[~df_order_items['order_id'].isin(lista_pedidos)]

# 3. Lista de pedidos en order_items
lista_pedidos_item = set(df_order_items['order_id'])

# 4. Listado de pedidos con productos que no tienen pedido
pedidos_item_invalidos = df_orders[~df_orders['order_id'].isin(lista_pedidos_item)]

# Resultados
print("Total de pedidos en ORDERS:", len(lista_pedidos))
print("Total de pedidos en ORDER_ITEMS:", len(lista_pedidos_item))
print("Pedidos de ORDERS_ITEMS que no están en ORDERS:", pedidos_sin_productos.shape[0])
print("Listado de productos con pedidos en ORDER que no están en ORDERS_ITEMS:", pedidos_item_invalidos.shape[0])

Total de pedidos en ORDERS: 96455
Total de pedidos en ORDER_ITEMS: 98666
Pedidos de ORDERS_ITEMS que no están en ORDERS: 2477
Listado de productos con pedidos en ORDER que no están en ORDERS_ITEMS: 0


### ORDER <-> ORDER_PAYMENTS

In [ ]:
# 1. Lista de pedidos en ORDER_PAYMENTS que no se encuentran en ORDERS (obtenido del punto anterior -> lista_pedidos)
order_payments_no_orders = df_order_payments[~df_order_payments['order_id'].isin(lista_pedidos)]

# 2. Lista de registros en ORDER_PAYMENTS
lista_order_payments = set(df_order_payments['order_id'])

# 3. Lista de registros de ORDERS que no están en ORDER_PAYMETS
lista_order_validos_en_order_payments = df_orders[~df_orders['order_id'].isin(lista_order_payments)]

# Resultado
print("Datos en ORDER_PAYMENTS que NO están en ORDERS:", len(order_payments_no_orders))
print("Órdenes en ORDER que no están en ORDER_PAYMENTS",len(lista_order_validos_en_order_payments))

Datos en ORDER_PAYMENTS que NO están en ORDERS: 3153
Órdenes en ORDER que no están en ORDER_PAYMENTS 1


### ORDER <-> ORDER_REVIEW

In [ ]:
# 1. Listado de ORDER_REVIEWS que no están en ORDERS
order_review_no_order = df_order_review[~df_order_review['order_id'].isin(lista_pedidos)]

# 2. Listado de ORDER_REVIEW
lista_order_review=set(df_order_review['order_id'])

# 3. Listado ORDERS que no están en ORDER REVIEWS
order_no_en_order_review = df_orders[~df_orders['order_id'].isin(lista_order_review)]

# Resultado
print("Listado de ORDER_REVIEW que no están en ORDER:", len(order_review_no_order))
print("Registros en ORDER que no están en ORDER_REVIEW:", len(order_no_en_order_review))

Listado de ORDER_REVIEW que no están en ORDER: 3008
Registros en ORDER que no están en ORDER_REVIEW: 0


### ORDER_ITEMS <-> ORDER_PAYMENTS

In [ ]:
# Dado los resultados anteriores se va a comparar la relación entre los registros de ORDER_ITEMS y ORDER_PAYMENTS
# 1. Registros de ORDER_ITEMS que no están en ORDER_PAYMENTS
order_items_no_en_order_paymets = df_order_items[~df_order_items['order_id'].isin(lista_order_payments)]

# 2. Registros de ORDER_PAYMENTS que no están en ORDER_ITEMS
order_payments_no_order_item=df_order_payments[~df_order_payments['order_id'].isin(lista_pedidos_item)]

# Resultado
print("Listado de ORDERS_ITEMS que no están en ORDER_PAYMENTS:",len(order_items_no_en_order_paymets))
print("Listado de ORDER_PAYMENTS que no están en ORDER_ITEMS:", len(order_payments_no_order_item))

Listado de ORDERS_ITEMS que no están en ORDER_PAYMENTS: 3
Listado de ORDER_PAYMENTS que no están en ORDER_ITEMS: 830


### PRODUCTS <-> ORDER_ITEMS

In [110]:
# 1. Lista de productos
lista_productos = set(df_products['product_id'])

# 2. Lista de productos en ORDER_ITEMS que no están en Products
products_no_order_items = df_order_items[~df_order_items['product_id'].isin(lista_productos)]

# 3. Lista de productos en ORDER_ITEMS
lista_productos_order_items=set(df_order_items['product_id'])

# 4. Lista de productos vendidos de la tabla ORDER_ITEMS que no están en la tabla PRODUCTS
productos_vendidos_no_products = df_products[~df_products['product_id'].isin(lista_productos_order_items)]

# Resultados
print("Productos vendidos que no están en la tabla PRODUCTS:", len(products_no_order_items))
print("Productos de la tabla PRODUCTS que no se hayan vendido:", len(productos_vendidos_no_products))

Productos vendidos que no están en la tabla PRODUCTS: 1628
Productos de la tabla PRODUCTS que no se hayan vendido: 0


### SELLER <-> ORDER_ITEMS

In [90]:
# 1. Lista de vendedores
lista_vendedores = set(df_sellers['seller_id'])

# 2. Lista de vendedores que no han vendido (no están en ORDER_ITEMS)
lista_vendedores_no_ventas = df_order_items[~df_order_items['seller_id'].isin(lista_vendedores)]

# 3. Lista de vendedores que han vendido
lista_vendedores_con_venta = set(df_order_items['seller_id'])

# 4. Lista de vendedores que han vendido que no están en tabla vendedores
vendedores_no_tabla_vendedor = df_sellers[~df_sellers['seller_id'].isin(lista_vendedores_con_venta)]


# Resultado
print("Lista de vendedores que no han vendido:", len(lista_vendedores_no_ventas))
print("Lista de vendedores con ventas que no están en tabla ventas:", len(vendedores_no_tabla_vendedor))

Lista de vendedores que no han vendido: 0
Lista de vendedores con ventas que no están en tabla ventas: 0


### SELLER <-> GEO_LOCATION

In [93]:
# 1. Lista de códigos postales de vendedores
lista_cp_vendedores = set(df_sellers['seller_zip_code_prefix'])

# 2. Listado de CP que no tienen vendedores
lista_cp_sin_vendedor = df_geolocation[~df_geolocation['geolocation_zip_code_prefix'].isin(lista_cp_vendedores)]

# 3. Listado de CP registrados
lista_cp = set(df_geolocation['geolocation_zip_code_prefix'])

# 4. CP de vendedores no registrados en CP de GEOLOCATION
lista_cp_vendedores_no_registrados = df_sellers[~df_sellers['seller_zip_code_prefix'].isin(lista_cp)] 

# Resultados
print("Lista de CP que no tienen vendedores:",len(lista_cp_sin_vendedor))
print("Lista de CP de vendedores que no están registrados en GEOLOCATION:", len(lista_cp_vendedores_no_registrados))

Lista de CP que no tienen vendedores: 16797
Lista de CP de vendedores que no están registrados en GEOLOCATION: 7


### CUSTOMER <-> GEO_LOCATION

In [105]:
# 1. Lista de cp de clientes
lista_cp_clientes = set(df_customers['customer_zip_code_prefix'])

# 2. Lista de cp de GEOLOCATION que no tienen clientes
lista_cp_geolocation_sin_clientes = df_geolocation[~df_geolocation['geolocation_zip_code_prefix'].isin(lista_cp_clientes)]

# 3. Listado de CP de clientes que no están en GEOLOCATION
lista_cp_clientes_no_geolocation = df_customers[~df_customers['customer_zip_code_prefix'].isin(lista_cp)]

# Resultado
print("Lista de CP de GEOLOCATION sin clientes:", len(lista_cp_geolocation_sin_clientes))
print("Lista de CP de clientes que no están en GEOLOCATION:", len(lista_cp_clientes_no_geolocation))

Lista de CP de GEOLOCATION sin clientes: 4178
Lista de CP de clientes que no están en GEOLOCATION: 278


___

### Conclusiones integridad tablas

#### Customers <-> Orders
* Hay 2986 clientes que no tienen órdenes de pedido
* Todos los clientes con órdenes de pedido tienen su correspondencia en la tabla CUSTOMERS (no hay clientes con pedidos que no estén registrados en la tabla customers)

#### Order <-> Order_Items
* La tabla Order_Items tiene 2477 órdenes que no están en la tabla Orders
* Todas las órdenes de la tabla Orders tiene su correspondencia en la tabla Order_Items (no hay órdenes de pedido que no tengan su correspondencia en order_item)

#### Order <-> Order_Payments
* La tabla Order_Payments tiene 3153 órdenes que no están recogidas en la tabla ORDERS.
* La tabla Orders tiene una órden de pedido que no tiene su correspondencia con la tabla de Order_Payment (o no se ha realizado el pago o no se ha contabilizado)

#### Order <-> Order_Review
* Hay 3008 órdenes de pedido que tiene review que no están referenciados en la tabla Orders
* Todas las órdenes de pedido recogidas en la tabla Orders tienen su review (no hay órdenes de la tabla orders que no tenga review -> Raro!!!)

#### Relación entre tablas hijas Order_Items y Order_Payments
* Hay tan solo 3 órdenes que no tiene pago referenciado en la tabla Order_Items
* Hay 830 órdenes de pago que no tienen orden de pedido en Order_Items (pedidos pagados sin productos?)
Esta relación inferior a los resultados obtenidos entre dichas tablas con la tabla Orders hace pensar que existen o existieron más pedidos de los que se reflejan en la tabla Orders, pero por alguna razón a la hora de extraer los datos no se incluyeron.

#### Products <-> Order_Items
* Hay 1628 productos vendidos en Order_Items que no se encuentran registrado en la tabla Products.
* Todos los productos de la tabla Products, tienen al menos una venta.

#### Seller <-> Order_Items
* Todos los vendedores tienen al menos una venta
* Todos los vendedores de la tabla Order_Items tiene su correspondencia en la tabla Seller.

#### Seller <-> Geolocation
* En la tabla Geolocation hay 16797 CP (Código Postal) que no tienen vendedores en dicha zona.
* Hay 7 vendedores cuyo CP no se encuentra en Geolocation.

#### Customer <-> Geolocation
* En la tabla Geolocation hay 4178 CP sin clientes
* En la tabla Customers hay 278 clientes con un CP que no se encuentra en Geolocation.

#### Gentileza de Gemini
### 📊 Informe de Auditoría de Integridad de Datos

Durante la fase de análisis y limpieza, se han auditado las relaciones entre las diferentes tablas del dataset para garantizar la fiabilidad del análisis posterior.

| Relación Analizada | Hallazgo Principal | Estado | Impacto Técnico / Conclusión |
| :--- | :--- | :---: | :--- |
| **Customers <-> Orders** | 2.986 clientes sin pedidos | ✅ | Clientes registrados que no completaron el embudo de ventas. |
| **Orders <-> Items** | 2.477 productos huérfanos | ⚠️ | Pérdida de cabeceras de pedido; inconsistencia en la exportación. |
| **Orders <-> Payments** | 3.153 pagos sin pedido | ❌ | Inconsistencia financiera; registros de pago sin pedido origen. |
| **Orders <-> Reviews** | 3.008 reseñas sin pedido | ⚠️ | Fallo de integridad; valoraciones de compras no registradas. |
| **Products <-> Items** | 1.628 ventas sin ficha | ⚠️ | Ventas registradas de productos que no existen en el catálogo. |
| **Sellers <-> Items** | **Sincronización Total** | **💎** | **Integridad perfecta. Relación de confianza del 100%.** |
| **Sellers <-> Geolocation** | 7 vendedores sin mapa | ✅ | Margen de error despreciable (< 0.1%) en geolocalización. |
| **Customers <-> Geolocation** | 278 clientes sin mapa | ✅ | Margen de error bajo (0.28%); aceptable en entornos reales. |